File: LeetCode/Python3/Problem 215.py
# LeetCode: Problem 215
# Jonathan Kosasih

#Find the kth largest element in an unsorted array.
#Note that it is the kth largest element in the sorted order, not the kth distinct element.
#
#Example 1:
#Input: [3,2,1,5,6,4] and k = 2
#Output: 5
#
#Example 2:
#Input: [3,2,3,1,2,4,5,5,6] and k = 4
#Output: 4
#------------------------------------------------------------------
import heapq
class Solution(object):
    # use heap data structure
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        heapq.heapify(nums)
        return heapq.nlargest(k, nums)[k-1]
    
    # just sort it in descending order and find the len(nums) - k index
    def findKthLargestAlt(self, nums, k):
        nums.sort(reverse = True)
        return nums[k-1]
    
sol = Solution()
nums = [3,2,1,5,6,4]
k = 2
print(str(sol.findKthLargest(nums, k)))
nums = [3,2,3,1,2,4,5,5,6]
k = 4
print(str(sol.findKthLargest(nums, k)))
print("--------------------------------")
nums = [3,2,1,5,6,4]
k = 2
print(str(sol.findKthLargestAlt(nums, k)))
nums = [3,2,3,1,2,4,5,5,6]
k = 4
print(str(sol.findKthLargestAlt(nums, k)))

File: LeetCode/C++/Problem 215.cpp
// LeetCode: Problem 215
// Jonathan Kosasih
/*
Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

// use max heap, largest number at top, pop k-1 times to find Kth largest
class Solution {
public:
	int findKthLargest(vector<int>& nums, int k) {
		priority_queue<int> pq(nums.begin(), nums.end());
		for (int i = 0; i < k - 1; i++) {
			pq.pop();
		}
		return pq.top();
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}